Sources/LLBCASFileTree/TSCCASFileSystem.swift

```diff
@@ -0,0 +1,144 @@
+        rootTree: LLBCASFileTree,
+        _ ctx: Context
+    ) {
+        self.db = db
+        self.client = LLBCASFSClient(db)
+        self.rootTree = rootTree
+        self.ctx = ctx
+    }
+
+    public func exists(_ path: AbsolutePath, followSymlink: Bool) -> Bool {
+        if path.isRoot {
+            return true
+        }
+        let result = try? self.rootTree.lookup(path: path, in: self.db, Context()).wait()
+        return result != nil
+    }
+
+    public func getDirectoryContents(_ path: AbsolutePath) throws -> [String] {
+        if path.isRoot {
+            return rootTree.files.map { $0.name }
+        }
+
+        let _result = try self.rootTree.lookup(path: path, in: self.db, ctx).wait()
+        guard let result = _result else { throw FileSystemError.noEntry }
+
+        // HACK: If this is a symlink, check if it points to a directory.
+        // Move this to LLBCASFileTree.lookup()
+        if result.info.type == .symlink, isDirectory(path) {
+            let symlinkContents = try readFileContents(path).cString
+            return try getDirectoryContents(path.parentDirectory.appending(RelativePath(symlinkContents)))
+        }
+
+        let entry = try self.client.load(result.id, ctx).wait()
+        guard let tree = entry.tree else { throw FileSystemError.notDirectory }
+        return tree.files.map{ $0.name }
+    }
+
+    public func isDirectory(_ path: AbsolutePath) -> Bool {
+        let fileType = self.fileType(of: path)
+        if fileType == .directory {
+            return true
+        }
+
+        // HACK: If this is a symlink, check if it points to a directory.
+        // Move this to LLBCASFileTree.lookup()
+        if fileType == .symlink {
+            guard let symlinkContents = try? readFileContents(path).cString else {
+                return false
+            }
+            return isDirectory(path.parentDirectory.appending(RelativePath(symlinkContents)))
+        }
+
+        return false
+    }
+
+    private func fileType(of path: AbsolutePath) -> LLBFileType? {
+        if path.isRoot { return .directory }
+        let fileType = try? self.rootTree.lookup(path: path, in: self.db, ctx).wait()
+        return fileType?.info.type
+    }
+
+    public func isFile(_ path: AbsolutePath) -> Bool {
+        let fileType = self.fileType(of: path)
+        return fileType == .plainFile || fileType == .executable
+    }
+
+    public func isExecutableFile(_ path: AbsolutePath) -> Bool {
+        fileType(of: path) == .executable
+    }
+
+    public func isSymlink(_ path: AbsolutePath) -> Bool {
+        fileType(of: path) == .symlink
+    }
+
+    public func readFileContents(_ path: AbsolutePath) throws -> ByteString {
+        if path.isRoot {
+            throw FileSystemError.ioError
+        }
+
+        let result = try rootTree.lookup(path: path, in: db, ctx).wait()
+        guard let id = result?.id else { throw FileSystemError.noEntry }
+
+        let entry = try client.load(id, ctx).wait()
```

Comment:
these 3 waits could be merged into a single wait using flatMaps, no? don't know what the overhead for waiting is, but might be better to wait a single time?